File: csharp-sdk/src/Cleanster/Api/PropertiesApi.cs
using System.Text.Json;
using Cleanster.Models;

namespace Cleanster.Api;

/// <summary>Manages physical cleaning locations.</summary>
public sealed class PropertiesApi
{
    private readonly ICleansterHttpClient _http;
    internal PropertiesApi(ICleansterHttpClient http) => _http = http;

    /// <summary>Return all properties, optionally filtered by service type.</summary>
    /// <param name="serviceId">Pass <see langword="null"/> to return all service types.</param>
    public async Task<ApiResponse<List<Property>>> ListPropertiesAsync(
        int? serviceId = null, CancellationToken ct = default)
    {
        var query = serviceId.HasValue
            ? new Dictionary<string, string> { ["serviceId"] = serviceId.Value.ToString() }
            : null;
        var root = await _http.GetAsync("/v1/properties", query, ct);
        return JsonHelper.ParseList<Property>(root);
    }

    /// <summary>Add a new property to the partner account.</summary>
    /// <param name="name">Property display name.</param>
    /// <param name="address">Street address.</param>
    /// <param name="city">City.</param>
    /// <param name="country">Country.</param>
    /// <param name="roomCount">Number of rooms.</param>
    /// <param name="bathroomCount">Number of bathrooms.</param>
    /// <param name="serviceId">Service type ID from <see cref="OtherApi.GetServicesAsync"/>.</param>
    /// <param name="state">Optional state/province.</param>
    /// <param name="zip">Optional postal/ZIP code.</param>
    /// <param name="timezone">Optional IANA timezone (e.g. "America/New_York").</param>
    /// <param name="note">Optional internal note visible to cleaners.</param>
    /// <param name="latitude">Optional GPS latitude.</param>
    /// <param name="longitude">Optional GPS longitude.</param>
    public async Task<ApiResponse<Property>> AddPropertyAsync(
        string name, string address, string city, string country,
        int roomCount, int bathroomCount, int serviceId,
        string? state = null, string? zip = null, string? timezone = null,
        string? note = null, double? latitude = null, double? longitude = null,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["name"]          = name,
            ["address"]       = address,
            ["city"]          = city,
            ["country"]       = country,
            ["roomCount"]     = roomCount,
            ["bathroomCount"] = bathroomCount,
            ["serviceId"]     = serviceId,
        };
        if (state     is not null) body["state"]     = state;
        if (zip       is not null) body["zip"]       = zip;
        if (timezone  is not null) body["timezone"]  = timezone;
        if (note      is not null) body["note"]      = note;
        if (latitude  is not null) body["latitude"]  = latitude;
        if (longitude is not null) body["longitude"] = longitude;
        var root = await _http.PostAsync("/v1/properties", body, ct);
        return JsonHelper.ParseSingle<Property>(root);
    }

    /// <summary>Return details of a specific property.</summary>
    public async Task<ApiResponse<Property>> GetPropertyAsync(int propertyId, CancellationToken ct = default)
    {
        var root = await _http.GetAsync($"/v1/properties/{propertyId}", ct: ct);
        return JsonHelper.ParseSingle<Property>(root);
    }

    /// <summary>Replace all fields of an existing property.</summary>
    /// <param name="state">Optional state/province.</param>
    /// <param name="zip">Optional postal/ZIP code.</param>
    /// <param name="timezone">Optional IANA timezone (e.g. "America/New_York").</param>
    /// <param name="note">Optional internal note visible to cleaners.</param>
    /// <param name="latitude">Optional GPS latitude.</param>
    /// <param name="longitude">Optional GPS longitude.</param>
    public async Task<ApiResponse<Property>> UpdatePropertyAsync(
        int propertyId, string name, string address, string city, string country,
        int roomCount, int bathroomCount, int serviceId,
        string? state = null, string? zip = null, string? timezone = null,
        string? note = null, double? latitude = null, double? longitude = null,
        CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?>
        {
            ["name"]          = name,
            ["address"]       = address,
            ["city"]          = city,
            ["country"]       = country,
            ["roomCount"]     = roomCount,
            ["bathroomCount"] = bathroomCount,
            ["serviceId"]     = serviceId,
        };
        if (state     is not null) body["state"]     = state;
        if (zip       is not null) body["zip"]       = zip;
        if (timezone  is not null) body["timezone"]  = timezone;
        if (note      is not null) body["note"]      = note;
        if (latitude  is not null) body["latitude"]  = latitude;
        if (longitude is not null) body["longitude"] = longitude;
        var root = await _http.PutAsync($"/v1/properties/{propertyId}", body, ct);
        return JsonHelper.ParseSingle<Property>(root);
    }

    /// <summary>Update freeform additional information for a property.</summary>
    public async Task<ApiResponse<JsonElement>> UpdateAdditionalInformationAsync(
        int propertyId, object data, CancellationToken ct = default)
    {
        var root = await _http.PutAsync($"/v1/properties/{propertyId}/additional-information", data, ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Toggle a property's active state.</summary>
    public async Task<ApiResponse<JsonElement>> EnableOrDisablePropertyAsync(
        int propertyId, bool enabled, CancellationToken ct = default)
    {
        var root = await _http.PostAsync($"/v1/properties/{propertyId}/enable-disable", new { enabled }, ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Permanently delete a property.</summary>
    public async Task<ApiResponse<JsonElement>> DeletePropertyAsync(int propertyId, CancellationToken ct = default)
    {
        var root = await _http.DeleteAsync($"/v1/properties/{propertyId}", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Return all cleaners currently assigned to a property.</summary>
    public async Task<ApiResponse<JsonElement>> GetPropertyCleanersAsync(int propertyId, CancellationToken ct = default)
    {
        var root = await _http.GetAsync($"/v1/properties/{propertyId}/cleaners", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Add a cleaner to a property's default cleaner pool.</summary>
    public async Task<ApiResponse<JsonElement>> AssignCleanerToPropertyAsync(
        int propertyId, int cleanerId, CancellationToken ct = default)
    {
        var root = await _http.PostAsync($"/v1/properties/{propertyId}/cleaners", new { cleanerId }, ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Remove a cleaner from a property's default cleaner pool.</summary>
    public async Task<ApiResponse<JsonElement>> UnassignCleanerFromPropertyAsync(
        int propertyId, int cleanerId, CancellationToken ct = default)
    {
        var root = await _http.DeleteAsync($"/v1/properties/{propertyId}/cleaners/{cleanerId}", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Set an iCal feed URL on a property for availability syncing.</summary>
    public async Task<ApiResponse<JsonElement>> AddICalLinkAsync(
        int propertyId, string icalLink, CancellationToken ct = default)
    {
        var root = await _http.PutAsync($"/v1/properties/{propertyId}/ical", new { icalLink }, ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Return the current iCal feed URL for a property.</summary>
    public async Task<ApiResponse<JsonElement>> GetICalLinkAsync(int propertyId, CancellationToken ct = default)
    {
        var root = await _http.GetAsync($"/v1/properties/{propertyId}/ical", ct: ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Remove the iCal feed URL from a property.</summary>
    public async Task<ApiResponse<JsonElement>> RemoveICalLinkAsync(
        int propertyId, string icalLink, CancellationToken ct = default)
    {
        var root = await _http.DeleteAsync($"/v1/properties/{propertyId}/ical", new { icalLink }, ct);
        return JsonHelper.ParseRaw(root);
    }

    /// <summary>Set the default checklist for a property.</summary>
    /// <param name="updateUpcomingBookings">If <see langword="true"/>, also applies to all future bookings at this property.</param>
    public async Task<ApiResponse<JsonElement>> SetDefaultChecklistAsync(
        int propertyId, int checklistId, bool updateUpcomingBookings = false, CancellationToken ct = default)
    {
        var path = $"/v1/properties/{propertyId}/checklist/{checklistId}?updateUpcomingBookings={updateUpcomingBookings.ToString().ToLower()}";
        var root = await _http.PutAsync(path, ct: ct);
        return JsonHelper.ParseRaw(root);
    }
}

File: android-sdk/src/main/kotlin/com/cleanster/android/api/PropertiesApi.kt
package com.cleanster.android.api

import com.cleanster.android.model.*
import retrofit2.Retrofit
import retrofit2.http.*

internal interface PropertiesService {
    @GET("v1/properties")
    suspend fun listProperties(@Query("serviceId") serviceId: Int? = null): ApiResponse<List<Property>>

    @POST("v1/properties")
    suspend fun addProperty(@Body body: CreatePropertyRequest): ApiResponse<Property>

    @GET("v1/properties/{propertyId}")
    suspend fun getProperty(@Path("propertyId") propertyId: Int): ApiResponse<Property>

    @PUT("v1/properties/{propertyId}")
    suspend fun updateProperty(
        @Path("propertyId") propertyId: Int,
        @Body body: CreatePropertyRequest,
    ): ApiResponse<Property>

    @DELETE("v1/properties/{propertyId}")
    suspend fun deleteProperty(@Path("propertyId") propertyId: Int): ApiResponse<Any>

    @PUT("v1/properties/{propertyId}/additional-information")
    suspend fun updateAdditionalInformation(
        @Path("propertyId") propertyId: Int,
        @Body body: Map<String, String>,
    ): ApiResponse<Any>

    @POST("v1/properties/{propertyId}/enable-disable")
    suspend fun enableOrDisableProperty(
        @Path("propertyId") propertyId: Int,
        @Body body: EnableDisableRequest,
    ): ApiResponse<Any>

    @GET("v1/properties/{propertyId}/cleaners")
    suspend fun getPropertyCleaners(@Path("propertyId") propertyId: Int): ApiResponse<List<Cleaner>>

    @POST("v1/properties/{propertyId}/cleaners")
    suspend fun addCleanerToProperty(
        @Path("propertyId") propertyId: Int,
        @Body body: AddCleanerToPropertyRequest,
    ): ApiResponse<Any>

    @DELETE("v1/properties/{propertyId}/cleaners/{cleanerId}")
    suspend fun removeCleanerFromProperty(
        @Path("propertyId") propertyId: Int,
        @Path("cleanerId") cleanerId: Int,
    ): ApiResponse<Any>

    @PUT("v1/properties/{propertyId}/ical")
    suspend fun setICalLink(
        @Path("propertyId") propertyId: Int,
        @Body body: ICalRequest,
    ): ApiResponse<Any>

    @GET("v1/properties/{propertyId}/ical")
    suspend fun getICalLink(@Path("propertyId") propertyId: Int): ApiResponse<Any>

    @HTTP(method = "DELETE", path = "v1/properties/{propertyId}/ical", hasBody = true)
    suspend fun deleteICalLink(
        @Path("propertyId") propertyId: Int,
        @Body body: ICalRequest,
    ): ApiResponse<Any>

    @PUT("v1/properties/{propertyId}/checklist/{checklistId}")
    suspend fun setDefaultChecklist(
        @Path("propertyId") propertyId: Int,
        @Path("checklistId") checklistId: Int,
        @Query("updateUpcomingBookings") updateUpcomingBookings: Boolean? = null,
    ): ApiResponse<Any>
}

class PropertiesApi(retrofit: Retrofit) {
    private val service = retrofit.create(PropertiesService::class.java)

    suspend fun listProperties(serviceId: Int? = null) =
        wrap { service.listProperties(serviceId) }

    suspend fun addProperty(request: CreatePropertyRequest) =
        wrap { service.addProperty(request) }

    suspend fun getProperty(propertyId: Int) =
        wrap { service.getProperty(propertyId) }

    suspend fun updateProperty(propertyId: Int, request: CreatePropertyRequest) =
        wrap { service.updateProperty(propertyId, request) }

    suspend fun deleteProperty(propertyId: Int) =
        wrap { service.deleteProperty(propertyId) }

    suspend fun updateAdditionalInformation(propertyId: Int, info: Map<String, String>) =
        wrap { service.updateAdditionalInformation(propertyId, info) }

    suspend fun enableOrDisableProperty(propertyId: Int, enabled: Boolean) =
        wrap { service.enableOrDisableProperty(propertyId, EnableDisableRequest(enabled)) }

    suspend fun getPropertyCleaners(propertyId: Int) =
        wrap { service.getPropertyCleaners(propertyId) }

    suspend fun addCleanerToProperty(propertyId: Int, cleanerId: Int) =
        wrap { service.addCleanerToProperty(propertyId, AddCleanerToPropertyRequest(cleanerId)) }

    suspend fun removeCleanerFromProperty(propertyId: Int, cleanerId: Int) =
        wrap { service.removeCleanerFromProperty(propertyId, cleanerId) }

    suspend fun setICalLink(propertyId: Int, icalLink: String) =
        wrap { service.setICalLink(propertyId, ICalRequest(icalLink)) }

    suspend fun getICalLink(propertyId: Int) =
        wrap { service.getICalLink(propertyId) }

    suspend fun deleteICalLink(propertyId: Int, icalLink: String) =
        wrap { service.deleteICalLink(propertyId, ICalRequest(icalLink)) }

    suspend fun setDefaultChecklist(
        propertyId: Int,
        checklistId: Int,
        updateUpcomingBookings: Boolean? = null,
    ) = wrap { service.setDefaultChecklist(propertyId, checklistId, updateUpcomingBookings) }
}
